src/wallet/wallet2.cpp

```diff
@@ -5021,40 +5021,83 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
   }
 }
 //----------------------------------------------------------------------------------------------------
-std::string wallet2::make_multisig(const epee::wipeable_string &password,
-  const std::vector<std::string> &initial_kex_msgs,
-  const std::uint32_t threshold)
+epee::misc_utils::auto_scope_leave_caller wallet2::decrypt_account_for_multisig(const epee::wipeable_string &password)
 {
   // decrypt account keys
+  // note: this conditional's clauses are old and undocumented
   epee::misc_utils::auto_scope_leave_caller keys_reencryptor;
   if (m_ask_password == AskPasswordToDecrypt && !m_unattended && !m_watch_only)
   {
     crypto::chacha_key chacha_key;
     crypto::generate_chacha_key(password.data(), password.size(), chacha_key, m_kdf_rounds);
-    m_account.encrypt_viewkey(chacha_key);
-    m_account.decrypt_keys(chacha_key);
+    this->decrypt_keys(chacha_key);
     keys_reencryptor = epee::misc_utils::create_scope_leave_handler(
-        [&, this, chacha_key]()
+        [this, chacha_key]()
         {
-          m_account.encrypt_keys(chacha_key);
-          m_account.decrypt_viewkey(chacha_key);
+          this->encrypt_keys(chacha_key);
         }
       );
   }
 
-  // create multisig account
-  multisig::multisig_account multisig_account{
-      multisig::get_multisig_blinded_secret_key(get_account().get_keys().m_spend_secret_key),
-      multisig::get_multisig_blinded_secret_key(get_account().get_keys().m_view_secret_key)
+  return keys_reencryptor;
+}
+//----------------------------------------------------------------------------------------------------
+void wallet2::get_uninitialized_multisig_account(multisig::multisig_account &account_out) const
```

Comment:
I'm not sure why an out parameter was chosen, but there's little benefit to it. The code generated isn't going to be more efficient; all modern ABIs have the caller pass a pointer via register that the callee uses to `std::move` or directly initialize into. So the primary difference is (typically) a worse interface for the caller.

Same thing for the function below since it appears to be overwriting `out` entirely too.

Reply:
Do you mean 'instead of returning an account'? Outparams are just a convention I decided to stick with for uniform interfaces (although I do break that occasionally, especially for trivial return types).

Reply:
Crap.

I've been pushing things towards `optional<T>` or `expect<T>` for various reasons, and away from the existing style of `bool` or `void` with out parameter (and towards a style where exceptions are _typically_ only for shutdown related events, which I kind-of violated in `wire::` serialization code).

Returning the type directly, especially when there is no `optional<T>` or `expect<T>`, is often easier on the caller code. Especially when the result is going directly into another function call, etc.

I thought perhaps you did this for performance reasons (typically that is what programmers cite), but this rationale is a bit different. If your happy with it, I'm not going to waste my time pushing back. Although it would be nice if seraphis used return-by-value where easily possible; (because as mentioned in last paragraph) return-by-value is often more composable.

Reply:
I'm actually skeptical about the value of ubiquitous function composition. Sometimes what you need for visual clarity is less code density, and sometimes what you need in order to realize 'I am writing spaghetti and need to decompose into more functions' is an excessive amount of lines of code.

Reply:
Also re: efficiency, sometimes an outparam is used to initialize an object in the caller and sometimes it is assigned to an object in the caller. For the assignment case I want to avoid excessive copies (although you seem to be implying this won't happen, so idk).

Reply:
> The code generated is typically close in both cases, but it also helps with bugs where an out parameter is aliasing another in parameter (waiting for Rust programmers to pile on here).

I suppose we are getting a bit off-topic, but this 'dangerous aliasing' is used for efficiency gains all over the crypto library.

```
crypto::secret_key a, b, c;
sc_add(a, a, b);  //a += b
sc_add(a, b, c);  //a = b + c
```

Where `sc_add()` is writing directly into the 32-byte outparam (the first parameter).

Reply:
> Also re: efficiency, sometimes an outparam is used to initialize an object in the caller and sometimes it is assigned to an object in the caller. For the assignment case I want to avoid excessive copies (although you seem to be implying this won't happen, so idk).

I'm going overkill on this explanation, but it is interesting.

If the out parameter is only being partially modified, it is clearly almost always more efficient than return-by-value. Otherwise, return-by-value is typically better because RVO/NRVO produces ASM (basically) identical to an out parameter that is uninitialized (by caller) and marked with the `restrict` keyword. Since usage of the `restrict` keyword is rare, and types are often initialized to some value immediately in C++, the "conventional wisdom" is to now prefer return-by-value in most situations.

When RVO/NRVO fails, moving is mandatory if possible, otherwise the fallback is copying. The move case is typically similar (in performance but not ASM) to out parameter or slightly slower. Might be slightly faster sometimes since `restrict` is rarely used on out parameters.

The fallback copying case basically always stinks, unless its a simple bit-copy (where move and copy are identical).

----
This confused me at first, so to explain the ASM - the caller just provides an uninitialized chunk of memory on the stack via register to the callee, which then does one single direct construction on the chunk of memory when RVO/NRVO is possible. This memory address can then be passed via register to another function - so `void handle(std::string); std::string a_string(); void run() {handle(a_string()); }` can produce ASM with 0 copies or moves and 1 initialization ... all the way back in C++03 days.

Compare to out parameter - `std::string` must be initialized by caller, and then mutated or overwritten by callee, etc.

Reply:
> When RVO/NRVO fails, moving is mandatory if possible, otherwise the fallback is copying. The move case is typically similar (in performance but not ASM) to out parameter or slightly slower. Might be slightly faster sometimes since restrict is rarely used on out parameters.

I may be underestimating compilers, but from what I understand RVO/NRVO doesn't help when you have composite types that are basically big blobs, which is what we are dealing with in the seraphis library.

For example I have this (64 bytes):

```
struct SpEnote final
{
    rct::key m_onetime_address;
    rct::key m_amount_commitment;
};
```

And this function:

```
void make_seraphis_enote_core(const rct::key &onetime_address,
    const crypto::secret_key &amount_blinding_factor,
    const rct::xmr_amount amount,
    SpEnote &enote_core_out)
{
    enote_core_out.m_onetime_address = onetime_address;
    enote_core_out.m_amount_commitment = rct::commit(amount, rct::sk2rct(amount_blinding_factor));
}
```

Both of the following cases require you to 1) instantiate an `SpEnote`, 2) copy-assign the members of the enote.

```
// 1) return value (internally make `SpEnote temp;` then return it to RVO initialize the `enote`)
const SpEnote enote{make_seraphis_enote_core(Ko, x, a)};

// 2) outparam
SpEnote enote;
make_seraphis_enote_core(Ko, x, a, enote);
```

With the outparam, I can reuse an existing `SpEnote` instance to make and consume multiple enotes in sequence (hypothetically) without copy-assigning it in the loop. [Here's](https://github.com/UkoeHB/monero/blob/c35686fcbbd17838cf23130c53495a8dd0e50502/src/seraphis/serialization_demo_utils.cpp#L157) an example of something like that (I did a search on my code and discovered I don't actually use this pattern very often).

I can also do partial-construction of objects without copy-assigning twice (found a bunch of examples in enote record recovery [for example](https://github.com/UkoeHB/monero/blob/c35686fcbbd17838cf23130c53495a8dd0e50502/src/seraphis/tx_enote_record_utils.cpp#L387)).

Reply:
> >    I can also do partial-construction of objects
> Meaning some of the memory is left uninitialized? This is a bad idea unless value/brace initialization is used everywhere for the structs or a default constructor initializes everything.

Sorry, I meant step-based member-wise initialization of an outparameter that is reused between function calls. With a return value you need to copy assign all the members then copy assign the return value to the temporary shuttle object.

> The standard recommendation is to always initialize variables on declaration in C++, which tips things towards return-by-value a bit.

I definitely agree that non-data types with some kind of 'behavior' must have a strict C++ interface with constructors and member functions. Doing that for data types where all your operations are data member writes and reads creates a kind of big mess of getters and setters. Maybe there is a way to do it that's clean, but I couldn't figure it out.

Reply:
I'm hihacking this PR (sorry), but for posterity:

> Sorry, I meant step-based member-wise initialization of an outparameter that is reused between function calls. With a return value you need to copy assign all the members then copy assign the return value to the temporary shuttle object.

I might be misunderstanding this statement, but if you meant something like:

```c++
struct blob { char value[32]; };
struct two_blobs { blob values[2]; };
blob return_blob();
two_blobs return_blobs() { return {return_blob(), return_blob()}; }
```

take that code snippet to gcc godbolt, you get:

```
return_blobs():
        push    rbx
        mov     rbx, rdi
        call    return_blob()
        lea     rdi, [rbx+32]
        call    return_blob()
        mov     rax, rbx
        pop     rbx
        ret
```

You should get similar results if `return_blobs()` used a named return variable. The callee just forwards the ("hidden") out address it received to the next function for initialization. It's identical to forwarding the address of the out parameter field/member. This of course, doesn't always work out exactly (so I'm _not_ proposing a change away from out variables).

Reply:
I mean something like this:

```
struct two_blobs { blob values[2]; };

bool try_get_blobs(const std::size_t i, two_blobs &blobs_out)
{
    if (!try_get_blob1(i, blobs_out[0])
        return false;

    get_blob2(blobs_out[1]);

    return true;
}

int main()
{
    two_blobs blobs_temp;
    for (std::size_t i{0}; i < n; ++i)
    {
        if (!try_get_blobs(i, blobs_temp)
            continue;

        do_something(blobs_temp);
    }
}
```